subprojects/gdk_rust/gdk_electrum/src/interface.rs

```diff
@@ -0,0 +1,946 @@
+            let all_utxos: Vec<&(BEOutPoint, UTXOInfo)> =
+                utxos.iter().filter(|(_, i)| i.asset == asset).collect();
+            let total_amount_utxos: u64 = all_utxos.iter().map(|(_, i)| i.value).sum();
+
+            let to_send = if asset == "btc" || Some(asset.to_string()) == self.network.policy_asset
+            {
+                let mut dummy_tx = BETransaction::new(self.network.id());
+                for utxo in all_utxos.iter() {
+                    dummy_tx.add_input(utxo.0.clone());
+                }
+                let estimated_fee = dummy_tx.estimated_fee(fee_rate, 1) + 3; // estimating 3 satoshi more as estimating less would later result in InsufficientFunds
+                total_amount_utxos
+                    .checked_sub(estimated_fee)
+                    .ok_or_else(|| Error::InsufficientFunds)?
+            } else {
+                total_amount_utxos
+            };
+
+            info!("send_all asset: {} to_send:{}", asset, to_send);
+
+            request.addressees[0].satoshi = to_send;
+        }
+
+        let mut tx = BETransaction::new(self.network.id());
+        // transaction is created in 3 steps:
+        // 1) adding requested outputs to tx outputs
+        // 2) adding enough utxso to inputs such that tx outputs and estimated fees are covered
+        // 3) adding change(s)
+
+        // STEP 1) add the outputs requested for this transactions
+        for out in request.addressees.iter() {
+            tx.add_output(&out.address, out.satoshi, out.asset_tag.clone())
+                .map_err(|_| Error::InvalidAddress)?;
+        }
+
+        // STEP 2) add utxos until tx outputs are covered (including fees) or fail
+        let mut used_utxo: HashSet<BEOutPoint> = HashSet::new();
+        loop {
+            let mut needs =
+                tx.needs(fee_rate, send_all, self.network.policy_asset.clone(), &wallet_data); // Vec<(asset_string, satoshi)  "policy asset" is last, in bitcoin asset_string="btc" and max 1 element
+            info!("needs: {:?}", needs);
+            if needs.is_empty() {
+                // SUCCESS tx doesn't need other inputs
+                break;
+            }
+            let current_need = needs.pop().unwrap(); // safe to unwrap just checked it's not empty
+
+            // taking only utxos of current asset considered, filters also utxos used in this loop
+            let mut asset_utxos: Vec<&(BEOutPoint, UTXOInfo)> = utxos
+                .iter()
+                .filter(|(o, i)| i.asset == current_need.asset && !used_utxo.contains(o))
+                .collect();
+
+            // sort by biggest utxo, random maybe another option, but it should be deterministically random (purely random breaks send_all algorithm)
+            asset_utxos.sort_by(|a, b| (a.1).value.cmp(&(b.1).value));
+            let utxo = asset_utxos.pop().ok_or(Error::InsufficientFunds)?;
+
+            // UTXO with same script must be spent together, even if it's another asset
```

Comment:
at the very least we have to always filter utxos smaller than dust (I think we can reasonably doing it at the source in the `fn utxos`, balance will reflect this change and I think it's correct to not consider outputs you are not going to spend).

More advanced economical consideration like not spending utxo in function of actual fee rate is desirable but it's more complicated and I think out of scope at the moment

For assets it's more complicated because there is no dust, for this reason, I think I will remove this logic for assets. Handling this would probably require coin blacklisting and again I think is out of scope at the moment

Reply:
dust removed from utxo in 9ca9c46a45e0a51e4b31f6fe74211326e1761892

Reply:
I could try implementing the economical considerations if you'd like.

The latest Bitcoin Optech newsletter happened to mention this very relevant SE question: [What's the most efficient way to create a raw transaction with a specific fee rate?](https://bitcoin.stackexchange.com/questions/98392/whats-the-most-efficient-way-to-create-a-raw-transaction-with-a-specific-fee-ra).

Reply:
That's interesting, I would love your help on this, if possible after merging this and gdk-electrum-2 if there is anything blocking